operators/nftopology/controllers/packagerevision_controller.go

```diff
@@ -0,0 +1,258 @@
+			pkgRev.Spec.Lifecycle)
+		return ctrl.Result{}, nil
+	}
+
+	var regionName, clusterName, nfInstanceName, topologyName, topoNamespace string
+	var createDeployed, found bool
+	pkgRevLabels := pkgRev.ObjectMeta.Labels
+	if regionName, found = pkgRevLabels["region"]; !found {
+		l.Info("region name not found from package")
+		return ctrl.Result{}, nil
+	}
+	if nfInstanceName, found = pkgRevLabels["nfinstance"]; !found {
+		l.Info("NF Instance name not found from package")
+		return ctrl.Result{}, nil
+	}
+	if topologyName, found = pkgRevLabels["nftopology"]; !found {
+		l.Info("NF Topology name not found from package")
+		return ctrl.Result{}, nil
+	}
+	if topoNamespace, found = pkgRevLabels["nftopology-namespace"]; !found {
+		l.Info("NF Topology namespace not found from package")
+		return ctrl.Result{}, nil
+	}
+	// repo name is the cluster name
+	clusterName = pkgRev.Spec.RepositoryName
+
+	// Step 2: search for NFTopology and NFDeployed
+	nfTopology := &nfreqv1alpha1.NFTopology{}
+	if err := r.Client.Get(ctx, client.ObjectKey{Namespace: topoNamespace, Name: topologyName}, nfTopology); err != nil {
+		l.Error(err, fmt.Sprintf("NFTopology object not found: %s\n", topologyName))
+		return ctrl.Result{}, nil
+	}
+
+	// TODO(s3wong): potentially, there would be multiple NFDeployedInstance CRs for an instance of
+	// NFInstance --- which means deploymentName should be unique
+	deploymentName := ConstructNFDeployedName(nfInstanceName, regionName, clusterName)
+
+	nfDeployed := &nfdeployv1alpha1.NFDeployed{}
+	if err := r.Client.Get(ctx, client.ObjectKey{Namespace: topoNamespace, Name: nfTopology.ObjectMeta.Name}, nfDeployed); err != nil {
+		createDeployed = true
+		nfDeployed.ObjectMeta.Name = nfTopology.ObjectMeta.Name
+		nfDeployed.ObjectMeta.Namespace = nfTopology.ObjectMeta.Namespace
+	}
+
+	// Search for corresponding NFInstance
+	var nfInstance *nfreqv1alpha1.NFInstance = nil
+	for _, nfInst := range nfTopology.Spec.NFInstances {
+		if nfInst.Name == nfInstanceName {
+			nfInstance = &nfInst
+			break
+		}
+	}
+	if nfInstance == nil {
+		err := errors.New(fmt.Sprintf("%s not found in NF Topology %s\n", nfInstanceName, topologyName))
+		l.Error(err, fmt.Sprintf("NF Instance %s not found in NF Topology %s\n", nfInstanceName, topologyName))
+		return ctrl.Result{}, err
+	}
+	// Get NFClass, cluster scope
+	className := nfInstance.NFTemplate.ClassName
+	var nfClass = &nfreqv1alpha1.NFClass{}
+	if err := r.Client.Get(ctx, client.ObjectKey{Name: className}, nfClass); err != nil {
+		l.Error(err, fmt.Sprintf("NFClass object not found: %s: %s\n", className, err.Error()))
+		return ctrl.Result{}, err
+	}
+	if err := BuildNFDeployed(deploymentName, topoNamespace, nfTopology, nfInstance, clusterName, nfClass.Spec.Vendor, nfClass.Spec.Version, nfDeployed); err != nil {
+		l.Error(err, fmt.Sprintf("Failed to build NFDeployed %s: %s\n", deploymentName, err.Error()))
+		return ctrl.Result{}, err
```

Comment:
For many of these errors, we should instead be setting the status conditions (probably a Ready condition) and using the message there. Then, the user can see the errors via `kubectl`. If we just return them like this, they may only end up in the log, which is hard to debug.

So, if you look at something like the PackageVariant controller, you will see that right in the beginning of Reconcile, we setup a `defer` call that updates the Status. Then, throughout the method, we set the conditions in various ways. When we exit, those conditions get written to the status in the api server.

Another thing you can do, but we have not in most places, is send events. I do this in the approval controller (in the nephio/controllers/pkg/reconcilers dir), because I don't "own" the resources I am messing with and so chose not to add a condition. I could have added a condition, that would be legit, but instead I decided to just send events. Those show up when you do a "describe" on the primary resource using kubectl.